rust/api/src/kv_store.rs

```diff
@@ -22,3 +25,602 @@ pub trait KvStore: Send + Sync {
+
+		let result = ctx.get_object("k1").await;
+		assert!(matches!(result, Err(VssError::NoSuchKeyError(_))));
+
+		Ok(())
+	}
+
+	async fn get_should_throw_no_such_key_exception_when_key_does_not_exist() -> Result<(), VssError>
+	{
+		let kv_store = Self::create_store().await;
+		let ctx = TestContext::new(&kv_store);
+
+		let result = ctx.get_object("non_existent_key").await;
+		assert!(matches!(result, Err(VssError::NoSuchKeyError(_))));
+
+		Ok(())
+	}
+	async fn get_should_return_correct_value_when_key_exists() -> Result<(), VssError> {
+		let kv_store = Self::create_store().await;
+		let ctx = TestContext::new(&kv_store);
+
+		ctx.put_objects(Some(0), vec![kv("k1", "k1v1", 0)]).await?;
+
+		let mut response = ctx.get_object("k1").await?;
+		assert_eq!(response.key, "k1");
+		assert_eq!(response.version, 1);
+		assert_eq!(response.value, Bytes::from("k1v1"));
+
+		let key_values = vec![kv("k1", "k1v2", 1), kv("k2", "k2v1", 0)];
+		ctx.put_objects(Some(1), key_values).await?;
+
+		response = ctx.get_object("k1").await?;
+		assert_eq!(response.key, "k1");
+		assert_eq!(response.version, 2);
+		assert_eq!(response.value, Bytes::from("k1v2"));
+
+		response = ctx.get_object("k2").await?;
+		assert_eq!(response.key, "k2");
+		assert_eq!(response.version, 1);
+		assert_eq!(response.value, Bytes::from("k2v1"));
+
+		let key_values = vec![kv("k2", "k2v2", 1), kv("k3", "k3v1", 0)];
+		ctx.put_objects(Some(2), key_values).await?;
+
+		response = ctx.get_object("k2").await?;
+		assert_eq!(response.key, "k2");
+		assert_eq!(response.version, 2);
+		assert_eq!(response.value, Bytes::from("k2v2"));
+
+		response = ctx.get_object("k3").await?;
+		assert_eq!(response.key, "k3");
+		assert_eq!(response.version, 1);
+		assert_eq!(response.value, Bytes::from("k3v1"));
+
+		Ok(())
+	}
+
+	async fn list_should_return_paginated_response() -> Result<(), VssError> {
+		let kv_store = Self::create_store().await;
+		let ctx = TestContext::new(&kv_store);
+
+		let total_kv_objects = 1000;
+		for i in 0..total_kv_objects {
+			ctx.put_objects(Some(i as i64), vec![kv(&format!("k{}", i), "k1v1", 0)]).await?;
+		}
+
+		ctx.put_objects(Some(1000), vec![kv("k1", "k1v2", 1)]).await?;
+		ctx.put_objects(Some(1001), vec![kv("k2", "k2v2", 1)]).await?;
+		ctx.put_objects(Some(1002), vec![kv("k2", "k2v3", 2)]).await?;
+
+		let mut previous_page: Option<ListKeyVersionsResponse> = None;
+		let mut all_key_versions: Vec<KeyValue> = Vec::new();
+
+		loop {
+			let current_page = if previous_page.is_none() {
+				let page = ctx.list(None, None, None).await?;
+				assert_eq!(page.global_version, Some(1003));
+				page
+			} else {
+				let next_page_token = previous_page.as_ref().unwrap().next_page_token.clone();
+				let page = ctx.list(next_page_token, None, None).await?;
+				assert!(page.global_version.is_none());
+				page
+			};
+
+			if current_page.key_versions.is_empty() {
+				break;
+			}
+
+			all_key_versions.extend(current_page.key_versions.clone());
```

Comment:
nit: can avoid the clone by storing just the `next_page_token` instead of the entire `ListKeyVersionsResponse`